### DISCUSSION TITLE
Binary Data getColour 'Attribute instanceColors is normalized' warning

### DISCUSSION THREAD

**Discussion by mcost45** [visgl/deck.gl#6119] [Q&A] [answered]
Hi, I've been making use of some binary data and I am getting the messages 'Attribute instanceColors / instanceLineColors is normalized'. Per the examples I've seen in documentation I've been supplying Uint8Arrays with rgba values, e.g. 255, 100, 50, 170, with an individual rgba value included for each vertex.

Does this message mean my Uint8Arrays are being normalised to Float32Arrays, and it would be faster if I inputted colours with Float32Arrays? Or that I may have used some wrong value elsewhere?

As an example. for a scatterplotLayer with 2 items, my getLineColor attribute is set to:
{value: [233, 30, 99, 180, 233, 30, 99, 180], size: 4}, with a Uint8Array. This will give the 'Attribute instanceLineColors is normalized' warning.

Thanks for any help!

**Comment by Pessimistress** [comment, marked as answer]
The warning is a little confusing. `normalized` is a flag passed to WebGLBuffer to cast byte values to the 0-1 range. The warning issued for color attributes if you provide an array that is not the default type -- in this case it happens to be `Uint8ClampedArray` -- AND do not explicitly set this flag. It is meant to warn you in case you provide a `Float32Array` with 0-255 values (WebGL buffer cannot normalize float values with the flag set either way). The following inputs work without warnings:

- `{value: <Uint8ClampedArray>, size}`
- `{value: <Uint8Array>, size, normalized: true}`
- `{value: <Float32Array>, size, normalized: false}`

We should make the following fixes:
- `{value: <Uint8Array>, size}` should not trigger any warning
- The warning should be better worded

**Reply by mcost45** [reply]
Thanks!!
